### PR TITLE
Include webserver in remaining pods under tnf testing to address lifecycle probes

### PR DESCRIPTION
**Pull request by ramperher**
- This affects to the images used by the example-cnf operators (excepting testpmd-container-app-mac, which is not currently used, so this will not be modified)
- For testpmd_image used currently, directly taken from dpdk image, we'll replace it by the image originally built from testpmd-container-app/cnfapp, which is reused again
- A webserver, based on golang (taking the skeleton from cnf-app-mac-operator), is included now in these images, and launched in background mode. Containers running in the same pod use different ports for the webserver (8095-8096)
- Liveness, readiness and startup probes have been defined in the corresponding pods
- Bump version to all the involved operators, container images and nfv-example-cnf-index
- Update GH actions to login in registry.redhat.io, since we need to download DPDK image from that repo

### CODE REVIEW

**Review by tonyskapunk — COMMENTED** [review]
First and foremost I suggest to hold on the change until knowing the redistribution limitations, if any.

---

Why the decision to use python for the webserver? Any requirement, limitation, etc?

If none, then you can avoid the rework by reusing the skeleton created in the cnf-app-mac-operator. It will also help to reduce "touching" too much the image, as only copying the binary and the entry-point.sh would get you to the same place. No need to install anything extra.

Same thoughts as in the cnf-app-mac-operator [here](https://github.com/openshift-kni/example-cnf/pull/35#issuecomment-1878757693). Specifically, if this is just to satisfy CNF, then is fine, but if we care about the purpose of those checks then this a good opportunity to add that value.

**Comment by ramperher** [comment]
> First and foremost I suggest to hold on the change until knowing the redistribution limitations, if any.
Sure, we'll ask next week about this to see what to do. Bad thing is that I fear that, if we're not really able to share an image based on dpdk-base-rhel8, we'll not be able to make modifications in the container using that image... so let's see

> Why the decision to use python for the webserver? Any requirement, limitation, etc?
> 
> If none, then you can avoid the rework by reusing the skeleton created in the cnf-app-mac-operator. It will also help to reduce "touching" too much the image, as only copying the binary and the entry-point.sh would get you to the same place. No need to install anything extra.
In fact, the skeleto for cnf-app-mac-operator is not really useful for the other cases since the logic for that operator is completely written in Go, whereas the others are written in Python. That's why I used Go for cnf-app-mac-operator, but Python for the others, just for simplicity and for, in fact, avoiding installing more binaries and software in the image that were just dedicated for this webserver.

> Same thoughts as in the cnf-app-mac-operator [here](https://github.com/openshift-kni/example-cnf/pull/35#issuecomment-1878757693). Specifically, if this is just to satisfy CNF, then is fine, but if we care about the purpose of those checks then this a good opportunity to add that value.
Yes, here's the same point, right now it's more an exercise to satisfy CNF requirements and also to confirm that the build of example-cnf is working fine, even modifying all the projects involved. Maybe I'd leave the discussion about the real meaning of these features for a second round of changes to improve this first implementation.

**Comment by ramperher** [comment]
The current status is the following:

- With the current code, the containers from loadbalancer are now working fine:

```
$ oc get pods -n example-cnf | grep loadbalancer
loadbalancer-ffccd4fd7-nrvd2                               2/2     Running            0             55s
$ oc exec -it loadbalancer-ffccd4fd7-nrvd2 -n example-cnf -c listener -- bash
[root@loadbalancer-ffccd4fd7-nrvd2 /]# curl 127.0.0.1:8095/healthz
query received to check liveness[root@loadbalancer-ffccd4fd7-nrvd2 /]# exit
exit
$ oc exec -it loadbalancer-ffccd4fd7-nrvd2 -n example-cnf -c loadbalancer -- bash                                                                                                                                                                                                                                               
[root@loadbalancer-ffccd4fd7-nrvd2 /]# curl 127.0.0.1:8096/healthz
query received to check liveness[root@loadbalancer-ffccd4fd7-nrvd2 /]# exit
exit                                                                                                                                                                                                                                                                                                                        
```

I still need to check in trex-config one but it should also work since it's similar.

However, for testpmd-app, I still need to figure out how to make it work for DPDK v4.6.3 (because I cannot install packages there; maybe I can use golang, which can be easily downloaded, and reuse the logic from cnf-app-mac-operator here) or for DPDK v4.14 (but this is not working currently)

**Comment by tonyskapunk** [comment]
> In fact, the skeleton for cnf-app-mac-operator is not really useful for the other cases since the logic for that operator is completely written in Go, whereas the others are written in Python. 

I'm sorry but I'm not following. What are you referring as "the others"? If you refer as the others to the webservers in here, then my question remains, why did you choose Python if in the end you're installing Python in the image? The same would've been possible by using a golang webserver and then **copyin** the binary into the image instead

> That's why I used Go for cnf-app-mac-operator, but Python for the others, just for simplicity and for, in fact, avoiding installing more binaries and software in the image that were just dedicated for this webserver.

I could understand that Python is simpler to leverage when nothing is created, but in this case the webserver in golang for cnf-app-mac is already available. In regards avoid installing software into the image, I found easier to copy the go binary rather than installing an interpreter and copying a script

> However, for testpmd-app, I still need to figure out how to make it work for DPDK v4.6.3 (because I cannot install packages there; maybe I can use golang, which can be easily downloaded, and reuse the logic from cnf-app-mac-operator here) or for DPDK v4.14 (but this is not working currently)

:upside_down_face:  This is the point that I'm trying to make for this change, why not using golang instead?

**Comment by ramperher** [comment]
check dallas ocp-4.14-vanilla example-cnf

**Comment by ramperher** [comment]
> * FAILURE https://www.distributed-ci.io/jobs/e1973225-59ea-4974-82f2-747ce5482492/jobStates
> * FAILURE https://www.distributed-ci.io/jobs/9008c68a-cb23-4cc9-abc7-b99f885c9088/jobStates
>   from [Include webserver in remaining pods under tnf testing to address lifecycle probes #38](https://github.com/openshift-kni/example-cnf/pull/38)

These tests are reporting a wrong result; in fact, the example-cnf installation worked: https://www.distributed-ci.io/jobs/9008c68a-cb23-4cc9-abc7-b99f885c9088, it's failing in a next step not related to this change and that has also appeared in the daily jobs this weekend.

All pods are up and running:

```
NAME                                                      READY   STATUS      RESTARTS        AGE     IP            NODE       NOMINATED NODE   READINESS GATES   LABELS
cnf-app-mac-operator-controller-manager-94c8b7c5f-fhfzh   1/1     Running     1 (4m32s ago)   9m11s   10.129.2.12   worker-1   <none>           <none>            control-plane=controller-manager,pod-template-hash=94c8b7c5f
job-trex-app-cqzzq                                        0/1     Completed   0               3m4s    10.129.2.16   worker-1   <none>           <none>            batch.kubernetes.io/controller-uid=eff1ff82-d012-4b0a-a746-12ca32399517,batch.kubernetes.io/job-name=job-trex-app,controller-uid=eff1ff82-d012-4b0a-a746-12ca32399517,example-cnf-type=pkt-gen-app,job-name=job-trex-app
loadbalancer-54d65f57b6-52k5c                             2/2     Running     0               5m22s   10.130.2.21   worker-3   <none>           <none>            example-cnf-type=lb-app,pod-template-hash=54d65f57b6
testpmd-app-76db89f94d-6jnb7                              1/1     Running     0               4m45s   10.128.2.37   worker-2   <none>           <none>            example-cnf-type=cnf-app,pod-template-hash=76db89f94d
testpmd-app-76db89f94d-7g9vl                              1/1     Running     0               4m45s   10.129.2.15   worker-1   <none>           <none>            example-cnf-type=cnf-app,pod-template-hash=76db89f94d
testpmd-lb-operator-controller-manager-cd99c4d4b-tshks    1/1     Running     0               8m9s    10.129.2.13   worker-1   <none>           <none>            control-plane=controller-manager,pod-template-hash=cd99c4d4b
testpmd-operator-controller-manager-599585987b-ttxgp      1/1     Running     0               6m57s   10.130.2.19   worker-3   <none>           <none>            control-plane=controller-manager,example-cnf-type=testpmd-operator,pod-template-hash=599585987b
trex-operator-controller-manager-6658c4f7c8-c9c8n         1/1     Running     0               5m57s   10.129.2.14   worker-1   <none>           <none>            control-plane=controller-manager,example-cnf-type=trex-operator,pod-template-hash=6658c4f7c8
trexconfig-5674785444-p26wp                               1/1     Running     0               3m54s   10.131.0.48   worker-0   <none>           <none>            example-cnf-type=pkt-gen,pod-template-hash=5674785444
```

I'll run a final test today to determine that everything is in place and that the targeted tnf tests are now passing.

**Comment by ramperher** [comment]
@tonyskapunk , finally I decided to follow your advice and move all the webserver's implementations to Go, but I'll answer your comments just to note here the decisions made.

> > In fact, the skeleton for cnf-app-mac-operator is not really useful for the other cases since the logic for that operator is completely written in Go, whereas the others are written in Python.
> 
> I'm sorry but I'm not following. What are you referring as "the others"? If you refer as the others to the webservers in here, then my question remains, why did you choose Python if in the end you're installing Python in the image? The same would've been possible by using a golang webserver and then **copyin** the binary into the image instead

What I was trying to say was the following:

- [cnf-app-mac-operator](https://github.com/openshift-kni/example-cnf/blob/main/cnf-app-mac-operator/Dockerfile#L2) is based in Go, so no new installations were needed to include the logic of the webserver; I directly included the code in Go.
- The other images, which are testpmd-container-app > [cnfapp](https://github.com/openshift-kni/example-cnf/blob/main/cnf-app-mac-operator/Dockerfile#L2), [testpmd](https://github.com/openshift-kni/example-cnf/blob/main/testpmd-container-app/testpmd/Dockerfile#L47) and [listener](https://github.com/openshift-kni/example-cnf/blob/main/testpmd-container-app/listener/Dockerfile#L1), and trex-container-app > [app](https://github.com/openshift-kni/example-cnf/blob/main/trex-container-app/app/Dockerfile#L1) and [server](https://github.com/openshift-kni/example-cnf/blob/main/trex-container-app/server/Dockerfile#L9), excepting the case of cnfapp (which was not used before and we're modifying it in this change), all the others are based on Python images or have Python installed. So, what I thought was that, maybe, it was easier in these cases to provide a Python script which implemented the webserver, only installing in the Dockerfile the modules that I needed, with pip. In case of using Go here, as you said (and that's what I've finally done), the Go binary was needed, which was not included in these images, whereas the Python binary was already in place.

That was the main reason why I used Python at the beginning; if these images were using Python, then it was just to continue using the same language. But okey, I finally moved to Go, so that all webservers implementations are based on that language, and also it's true that this solved my problem with cnfapp, where I was not able to download any package with dnf/yum because of subscription issues, and the Go binary, I was able to provide it with the ADD clause from Dockerfile.

> > That's why I used Go for cnf-app-mac-operator, but Python for the others, just for simplicity and for, in fact, avoiding installing more binaries and software in the image that were just dedicated for this webserver.
> 
> I could understand that Python is simpler to leverage when nothing is created, but in this case the webserver in golang for cnf-app-mac is already available. In regards avoid installing software into the image, I found easier to copy the go binary rather than installing an interpreter and copying a script
> 
> > However, for testpmd-app, I still need to figure out how to make it work for DPDK v4.6.3 (because I cannot install packages there; maybe I can use golang, which can be easily downloaded, and reuse the logic from cnf-app-mac-operator here) or for DPDK v4.14 (but this is not working currently)
> 
> 🙃 This is the point that I'm trying to make for this change, why not using golang instead?

Already answered above.

**Comment by ramperher** [comment]
In the last debug job I'm launching (I've run a complete example-cnf installation), the targeted tests are now passing!

```
$ cat /tmp/dci_logs.mrb5y63b/cnf-certification-tests_junit.xml | grep probe                                                                                                               
          <testcase name="[It] lifecycle lifecycle-readiness-probe [telco, lifecycle-readiness-probe, lifecycle]" classname="CNF Certification Test Suite" status="passed" time="0.000147667">  
          <testcase name="[It] lifecycle lifecycle-liveness-probe [telco, lifecycle-liveness-probe, lifecycle]" classname="CNF Certification Test Suite" status="passed" time="0.000145401">                       
          <testcase name="[It] lifecycle lifecycle-startup-probe [telco, lifecycle-startup-probe, lifecycle]" classname="CNF Certification Test Suite" status="passed" time="0.00013513">
```

I'll wait for the end of the execution to confirm everything is fine, but this is a good first result.

**Comment by ramperher** [comment]
The complete example-cnf deployment is passing:

- tnf probes tests are now passing
- testpmd-container-app/cnfapp image is correctly used
- other tests, like preflight and opcap, are also executed correctly

Change is ready for review

File: testpmd-container-app/cnfapp/Dockerfile [hunk 1]
```diff
@@ -1,20 +1,32 @@
...
-        --enablerepo rhel-8-for-x86_64-appstream-rpms --setopt=tsflags=nodocs \
-        --security --sec-severity=Important --sec-severity=Critical
 
-USER default
+# Install Go binary
+ARG GO_VERSION=go1.21.6
+ADD https://go.dev/dl/$GO_VERSION.linux-amd64.tar.gz /usr/local
+RUN tar -C /usr/local -xzf /usr/local/$GO_VERSION.linux-amd64.tar.gz
+RUN rm /usr/local/$GO_VERSION.linux-amd64.tar.gz
+ENV PATH="$PATH:/usr/local/go/bin"
```
Comment (tonyskapunk) [inline testpmd-container-app/cnfapp/Dockerfile:22]: There's no need to install go in here, use two container instead, one for build and one to ship, only copy the binary in the last one (to ship), like this:

```Dockerfile
FROM docker.io/library/golang:1.21 as builder

WORKDIR /scripts
COPY scripts/webserver.go .
RUN go mod init webserver.go
RUN GOOS=linux CGO_ENABLED=0 go build -a -o webserver .

# This image needs confirmation :warning:
FROM registry.redhat.io/openshift4/dpdk-base-rhel8:v4.6.3-8

-- snip --

COPY chmod u+x scripts/entrypoint.sh  /usr/local/bin/
COPY chmod u+x --from=builder /scripts/webserver /usr/local/bin/

ENTRYPOINT ["/usr/local/bin/entrypoint.sh"]
```

This way you don't need to install anything as I mentioned previously, just copy the binary.

**Review by tonyskapunk — CHANGES_REQUESTED** [review]
Glad tnat we are making progress here, please take a look to my comment on the containerfile.

Also consider avoiding repetition if the same code will be used, use a single source and reuse it.

**Comment by ramperher** [comment]
check dallas ocp-4.15-vanilla example-cnf

File: testpmd-container-app/cnfapp/Dockerfile (hunk 1 above)
Comment (ramperher) [reply testpmd-container-app/cnfapp/Dockerfile]: I think I have done this correctly, please confirn

File: testpmd-container-app/cnfapp/scripts/entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+set -ex
+
+# Start lifecycle webserver in background
+go run /usr/local/bin/webserver.go &
```
[the remark this answers is not in the document]
Comment (ramperher) [reply testpmd-container-app/cnfapp/scripts/entrypoint.sh]: Done

**Comment by ramperher** [comment]
> Glad tnat we are making progress here, please take a look to my comment on the containerfile.
> 
> Also consider avoiding repetition if the same code will be used, use a single source and reuse it.

Thanks @tonyskapunk ! Regarding the repetition, I tried by using a symbolic link to a file I have placed in a new folder called common-scripts, but the build complained about that and did not work, I have used a hard link instead. Do you have a better way of handling this?

I will run a job in the meantime to see how it goes with the current build

**Comment by ramperher** [comment]
don't know what's going on with cluster5... I'll retest tomorrow, in the meantime the change can be reviewed

**Comment by ramperher** [comment]
check dallas ocp-4.13-vanilla example-cnf

File: trex-container-app/server/Dockerfile [hunk 3]
```diff
@@ -22,4 +29,9 @@ ENV PYTHONPATH="/opt/trex/trex-core/scripts/automation/trex_control_plane/intera
 WORKDIR /opt/trex/trex-core/scripts
 ENV TREX_DIR="/opt/trex/trex-core/scripts"
 ENV TRAFFICGEN_DIR="/opt/trafficgen"
-COPY scripts /usr/local/bin/
+
+# Copy scripts
+COPY scripts /usr/local/bin
+COPY --from=build /scripts/webserver /usr/local/bin/
+
+RUN chmod +x /usr/local/bin/webserver
```
Comment (tonyskapunk) [inline trex-container-app/server/Dockerfile:37]: Avoid copying anything that is not needed in the image that is inside the scripts, e.g. the webserver.go

```suggestion
COPY --chmod=550 --from=build /scripts/webserver /usr/local/bin/
```

File: testpmd-container-app/cnfapp/Dockerfile [hunk 4]
```diff
@@ -1,20 +1,33 @@
...
-        --security --sec-severity=Important --sec-severity=Critical
 
-USER default
+# Copy scripts
+COPY scripts /usr/local/bin
+COPY --from=build /scripts/webserver /usr/local/bin/
+
+# Update permissions
+RUN chmod +x /usr/local/bin/webserver
+RUN chmod +x /usr/local/bin/entrypoint.sh
```
Comment (tonyskapunk) [inline testpmd-container-app/cnfapp/Dockerfile:30]: Avoid copying anything that is not needed in the image that is inside the scripts, e.g. the webserver.go

```suggestion
COPY --chmod=550 scripts/entrypoint.sh /usr/local/bin/
COPY --chmod=550 --from=build /scripts/webserver /usr/local/bin/
```

File: testpmd-container-app/listener/Dockerfile [hunk 5]
```diff
@@ -1,5 +1,19 @@
...
 FROM registry.access.redhat.com/ubi8/python-39:latest
 USER root
 RUN pip3 install kubernetes
+
+# Copy scripts
 COPY scripts /usr/local/bin
+COPY --from=build /scripts/webserver /usr/local/bin/
+
+# Update permissions
+RUN chmod +x /usr/local/bin/webserver
```
Comment (tonyskapunk) [inline testpmd-container-app/listener/Dockerfile:17]: Avoid copying anything that is not needed in the image that is inside the scripts, e.g. the webserver.go

```suggestion
COPY --chmod=550 --from=build /scripts/webserver /usr/local/bin/
```

File: testpmd-container-app/testpmd/Dockerfile [hunk 6]
```diff
@@ -51,5 +59,12 @@ RUN dnf install -y \
 
 # copy testpmd runtime cmdline file
 COPY testpmd-runtime-cmds.txt /root/testpmd-runtime-cmds.txt
+
+# Copy scripts
 COPY scripts /usr/local/bin
+COPY --from=build2 /scripts/webserver /usr/local/bin/
+
+# Update permissions
+RUN chmod +x /usr/local/bin/webserver
```
Comment (tonyskapunk) [inline testpmd-container-app/testpmd/Dockerfile:68]: Avoid copying anything that is not needed in the image that is inside the scripts, e.g. the webserver.go

```suggestion
COPY --chmod=550 --from=build2 /scripts/webserver /usr/local/bin/
```

File: trex-container-app/app/Dockerfile [hunk 7]
```diff
@@ -15,6 +22,12 @@ ENV PYTHONPATH="/opt/trex/trex-core/scripts/automation/trex_control_plane/intera
 ENV TREX_DIR="/opt/trex/trex-core/scripts"
 ENV TRAFFICGEN_DIR="/opt/trafficgen"
 
+# Copy scripts
 COPY scripts /usr/local/bin/
+COPY --from=build /scripts/webserver /usr/local/bin/
+
+# Update permissions
+RUN chmod +x /usr/local/bin/webserver
```
Comment (tonyskapunk) [inline trex-container-app/app/Dockerfile:30]: Avoid copying anything that is not needed in the image that is inside the scripts, e.g. the webserver.go

```suggestion
COPY --chmod=550 --from=build /scripts/webserver /usr/local/bin/
```

**Review by tonyskapunk — CHANGES_REQUESTED** [review]
> Thanks @tonyskapunk ! Regarding the repetition, I tried by using a symbolic link to a file I have placed in a new folder called common-scripts, but the build complained about that and did not work, I have used a hard link instead. Do you have a better way of handling this?

I don´t  know if hard links work in git, likely not, but you can take the script out of all this directories and put it under ./utils or something and use it from there

> One option I was thinking is to move webserver.go to a personal repo under my github account, making it public, and add it to the golang build image, don’t know if this can satisfy the requirement

Certainly Is an option but not sure if is worth the effort, here remains local. And I think it will eventually change to satisfy more than just replying to http requests. My 2¢.

File: testpmd-container-app/cnfapp/scripts/webserver.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
+/*
+
```
[the remark this answers is not in the document]
Comment (ramperher) [reply testpmd-container-app/cnfapp/scripts/webserver.go:2]: added

File: trex-container-app/server/Dockerfile (hunk 3 above)
Comment (ramperher) [reply trex-container-app/server/Dockerfile]: done

**Comment by ramperher** [comment]
> > Thanks @tonyskapunk ! Regarding the repetition, I tried by using a symbolic link to a file I have placed in a new folder called common-scripts, but the build complained about that and did not work, I have used a hard link instead. Do you have a better way of handling this?
> 
> I don´t know if hard links work in git, likely not, but you can take the script out of all this directories and put it under ./utils or something and use it from there
> 
> > One option I was thinking is to move webserver.go to a personal repo under my github account, making it public, and add it to the golang build image, don’t know if this can satisfy the requirement
> 
> Certainly Is an option but not sure if is worth the effort, here remains local. And I think it will eventually change to satisfy more than just replying to http requests. My 2¢.

The `utils` directory is a good idea, then I'm able to differentiate between scripts and utils in this way.

I finally decided to do the following:

- Create `utils/webserver.go` for each image requiring this feature
- Create `utils/webserver.go` in the top level of the repository, as a main source of code
- Include, in the Makefiles that build the container images, an instruction that copies the webserver.go from the top level to each image folder, so that all the images use the same source code as base in the automation (but, for making it compatible with a local copy, that's why I've maintained in the repo a local copy on each image)

I've also applied your suggestions, building the image now and I'll run a final test to see how it goes. I'll re-request your review again, thanks for all the effort here :)

**Comment by ramperher** [comment]
Last job passed and we can see the fix for lifecycle probe tnf tests is working, since the tests are passing. This is ready for review again. CC @tonyskapunk

**Review by tonyskapunk — APPROVED** [review]
LGTM

As discussed offline, please create a card to track the findings on  registry.redhat.io/openshift4/dpdk-base-rhel8:v4.6.3-8

Thanks for all the work!

**Pull request merged**